controller/src/main/java/io/pravega/controller/server/security/auth/AuthorizationResourceImpl.java

```diff
@@ -64,4 +67,15 @@ public String ofKeyValueTableInScope(String scopeName, String keyValueTableName)
         Exceptions.checkNotNullOrEmpty(keyValueTableName, "keyValueTableName");
         return String.format("%s/%s:%s", ofScope(scopeName), TAG_KEYVALUETABLE, keyValueTableName);
     }
+
+    @Override
+    public String ofInternalStream(String scopeName, @NonNull String streamName) {
+        if (streamName.startsWith(NameUtils.READER_GROUP_STREAM_PREFIX)) {
```

Comment:
Reader Group synchronizer and Mark streams are internal streams, but they have one very important difference. The former is written directly by clients (readers) while the latter is written by the controller. For the Mark stream, what we need to authorize is that controller tracks the requests to note time from the writer, and the writer is already expected to have write access. A reader will only be reading to the Mark stream, not writing to it.

Reply:
For watermark streams we authorize against the corresponding stream: 

https://github.com/pravega/pravega/blob/11daa62619d48b59369586c0aa8444d7df75359e/controller/src/main/java/io/pravega/controller/server/security/auth/AuthorizationResourceImpl.java#L75-L76

Does that look OK?

Reply:
My point is that in principle no client needs to write to the mark stream, only the controller writes to the mark stream. We shouldn't be granting the client authorization to write to that stream if it doesn't need it.

Reply:
The segment store also authorized the controller's requests, so it must have permissions to write to that watermark stream. Since the client doesn't invoke the operation, it wouldn't be authorized for the mark stream or issued a delegation token containing the authorization. Does that address your question, or am I missing something?

Reply:
I'm not sure it addresses my question. The controller appends to the mark stream, and I think you are saying that the controller appending to the mark stream also requires authorization. This part makes sense. 

The next question I have is how the controller is authorized to append to the mark stream. The controller consolidates in a single stream cut the marks from a group of writers, so it can't be transitively authorized with the token of a single writer. In fact, I'm wondering if its authorization should have anything to do with the authorization of the writers as a controller instance should be able to perform operations against the segment store as long as it can demonstrate that it is a legitimate controller. If a controller has been authenticated as a legitimate controller, then it can be authorized as such to perform operations against the segment store. What am I missing?

Reply:
I don't know exactly what/how appends are facilitated by the controller and made by the segment store, so I don't know the answers to all the questions above. @shiveshr / @tkaitchuck could you please chime in?

Based on my experiments, I had seen that the reads from a stream'd fail if the user didn't have permissions on the corresponding mark stream. (See references `_MARK` stream in [this PDP](https://github.com/pravega/pravega/wiki/PDP-46:-Read-Only-Permissions-For-Reading-Data#motivation). Since mark streams are an internal construct, I had modified the code to authorize against the corresponding stream in this PR, based on my discussion with @shiveshr. My understanding is that for a stream `myscope/mystream` we create a corresponding mark stream `myscope/_MARKmystream`.

Reply:
@fpj is correct. 

writers need to be authorized for read-update on base stream to be able to call notetime. 
writers do not request for a delegation token for calling notetime as they do not write directly to mark stream. 

Controller uses the "master" delegation token for all writes to all metadata tables and mark streams. So it has permissions by default for writing to any internal stream. Note: for table calls it creates the master delegation token. for client calls (e.g. request stream reader and writer, mark's revisioned stream client, it uses the ClientConfig which loads the creds from the environment which gives it the controller creds for talking to segment store).

Reply:
@shiveshr yes, if the client requests for a delegation token (via `getDelegationToken` call) with `READ_UPDATE` permissions for a mark stream, we validate that the client has `READ_UPDATE` permission on the corresponding stream and return a delegation token enabling the client to read from/write to that mark stream. We aren't preventing that from happening as of now. 

If clients have no business requesting for delegation tokens for mark streams, I can add logic to prevent that from happening. What do you say?

Reply:
prevent them from requesting read_update. 
readers still need "read" permissions on it to read from mark stream.

Reply:
I've made the change in the latest commit that I just pushed. Please take a look. Here's what it does: 

1. If the client calls `getDelegationToken` for read permissions on mark streams, it authorizes the user and returns one. 
2. If the client calls `getDelegationToken` for write permissions on mark streams, it rejects the request with an authorization exception. 
3. If the client calls `getDelegationToken` for markstreams, without specifying the access operation, it authorizes the client for reads and returns a delegation token with read permissions.

Reply:
There is one point I'd like to clarify, see this:

https://github.com/pravega/pravega/pull/5231#discussion_r522426709